qcs/src/executable.rs

```diff
@@ -362,64 +347,141 @@ impl<'execution> Executable<'_, 'execution> {
     /// [quilc]: https://github.com/quil-lang/quilc
     pub async fn execute_on_qpu(&mut self, quantum_processor_id: &'execution str) -> ExecuteResult {
         let mut qpu = self.qpu_for_id(quantum_processor_id).await?;
-        let mut config = self.get_config().await;
+        let mut config = self.get_config().await?;
 
         let response = match qpu.run(&self.params, self.get_readouts(), &config).await {
-            Ok(result) => Ok(result),
-            Err(qpu::Error::Qcs {
-                retry_after: None, ..
-            }) => {
-                // If retry_after is set, don't retry now
-                config = self
-                    .refresh_config()
-                    .await
-                    .wrap_err("When refreshing authentication token")?;
+            Ok(response) => Ok(response),
+            Err(ExecutionError::Unauthorized) => {
+                config = self.refresh_config().await?;
                 qpu.run(&self.params, self.get_readouts(), &config).await
             }
-            err => err,
+            Err(err) => Err(err),
         };
 
-        self.qpu = Some(qpu);
-        match response {
-            Ok(result) => Ok(result),
-            Err(qpu::Error::Qcs {
-                source,
-                retry_after,
-            }) => {
-                match retry_after {
-                    Some(duration) => {
-                        // retry_after could mean maintenance which requires recompile
-                        self.qpu = None;
-                        Err(Error::Retry {
-                            source,
-                            after: duration,
-                        })
-                    }
-                    None => Err(source.into()),
-                }
-            }
-            Err(quil_err @ qpu::Error::Quil { .. }) => Err(Error::from(
-                eyre!(quil_err).wrap_err("When compiling for the QPU"),
-            )),
-        }
+        self.qpu = if let Err(ExecutionError::QpuUnavailable(_)) = response {
+            // Could mean maintenance which requires recompile
+            None
```

Comment:
Is this the only error that should cause `self.qpu = None` ? I assume the others would cause the request to fail earlier?

Reply:
Other failures do not necessarily need to void the compilation cache (which is what setting this to `None` does). The `QpuUnavailable` error is the only one which signifies planned maintenance—which could mean the QPU settings changed.